[PR-34] TeamMain 화면 구성 및 도메인 영역 작업

ohjuntaek opened the pull request:
resolved : #33 

https://app.zeplin.io/project/5e86c71cde586b485f3b4e14/screen/5e9a970686460978eeb73c30

- [x] 커밋 1. 더미 데이터로 recyclerView(TeamFeed) 제외하고 화면 및 로직 구성
- [x] 커밋 2. 더미 데이터로 recyclerView 화면 및 로직 구성
- [x] 커밋 3. 백엔드 API 연결 작업
- [x] 커밋 4. 전체 코드 리뷰 반영 및 merge conflict를 해소를 위한 수정 작업
  - shimmer 추가
-----------

- [ ] 커밋 5. FEAT : 리더 사용자 이외 사용자 고려하여 구현
- [ ] 커밋 6. REFACTOR : Spannable 사용 
=> 이 부분은 다른 쪽 화면 작업 끝나고 진행할게요.

화면 스크린샷

![image](https://user-images.githubusercontent.com/17001524/81626141-cfe09180-9435-11ea-8078-2ee3d9893dab.png)

![image](https://user-images.githubusercontent.com/17001524/81626153-d4a54580-9435-11ea-913e-b6445dc73d3b.png)

ohjuntaek committed:
FEAT : TeamMain 화면 구성 및 도메인 영역 작업

LeeOhHyung commented on app/src/main/java/kr/yapp/teamplay/teammain/data/TeamMainItem.kt:
```diff
@@ -0,0 +1,20 @@
+ * TODO : API 정의 후 업데이트
+ * TODO : 동호회 게시판(Feed) 추가
+ */
+data class TeamMainItem(
+    @SerializedName("tag") val tag: String,
+    @SerializedName("name") val name: String,
+    @SerializedName("address") val address: String,
+    @SerializedName("create_date") val createDate: String,
+    @SerializedName("member_count") val memberCount: String
+)
```
해당 Data class의 멤버변수는 서버쪽 정의를 참고하신것이죠??
PR본문에 서버API 관련 문서를 링크해주시면 리뷰하는데 더 도움이 될것같아요..

ohjuntaek replied:
아직 api가 나오지 않은걸로 알고 있어 화면 참고해서 진행 중이고 추후 api 반영 예정입니다.

LeeOhHyung commented on app/src/main/java/kr/yapp/teamplay/teammain/domain/TeamMainRepositoryImpl.kt:
```diff
@@ -0,0 +1,21 @@
+import kr.yapp.teamplay.teammain.data.TeamMainItem
+import kr.yapp.teamplay.teammain.data.TeamMainRepository
+
+class TeamMainRepositoryImpl(
+    private val TeamMainService: Any = ""
+) : TeamMainRepository {
```
`TeamMainService `의 타입을 Any로 선언하신 이유가 궁금합니다.

ohjuntaek replied:
더미데이터를 넣기 위해 적용했고 추후에 서버 통신하면 레트로핏 타입으로 변경예정입니다.

LeeOhHyung commented on app/src/main/res/drawable/team_main_tv_top.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+
+    <solid android:color="#828282" />
+    <corners android:radius="13dp" />
+
+</shape>
```
`radius` 12dp 또는 14dp인데 혹시 오타 아닐까 예상해봅니다..

ohjuntaek replied:
제플린 보고 한거 같은데 확인해보겠습니다.

LeeOhHyung reviewed (approved):
고생하셨습니다~~

ckddlr0610 commented:
수고하셨습니다! xml이나 Repository의 변수들의 네이밍이 좋은 것 같아요

teddy-choe commented on app/src/main/res/layout/fragment_team_main.xml:
```diff
@@ -0,0 +1,187 @@
+        android:paddingEnd="15dp">
+
+        <View
+            android:id="@+id/team_main_bg_top"
```
전반적으로 변수들의 네이밍이 통일성있게 좋아보입니다

teddy-choe reviewed (approved):
수고하셨습니다! 전반적으로 변수 네이밍이 좋아서 많이 참고할 것 같습니다

ohjuntaek committed:
FEAT : TeamMainFeed 화면 구성 및 도메인 작업

teddy-choe commented on app/src/main/java/kr/yapp/teamplay/teammain/data/NoticeItem.kt:
```diff
@@ -0,0 +1,9 @@
+    val title: String = "3월 2주 공지입니다",
+    val createTime: String = "2020.3.9. 16:00",
+    val content: String = "이번 주 토 3월 15일 오후 4시부터 7시 남양주 농구 동호회 팀 열정과 3대3 반코트 경기가 있습니다. 3대3 반코트 게임을 진행할 예정입니다. 게임 후 식사 및 간단하게 한 잔 있어요!"
+) : Serializable
```
Serializable을 extends하신것은 `@SerializedName` 대신에 사용하기 위한 것인가요??? Serializable이 어떤 역할을 하는 것인지 궁금합니다

ohjuntaek replied:
자바에서 직렬화라는 개념이 있습니다. 데이터를 전송하기 위해 jvm의 힙, 스택, 상수 영역에 있는 객체안의 데이터를 모아서 바이트로 변환하는 것을 의미한다고 알고있어요. 
안드에서는 주로 프래그먼트, 액티비티 사이에 객체를 전달할 때(번들 or 인텐드로 전달하기 때문에) 직렬화가 필요합니다.
사실 저 객체에 아직까진 직렬화가 필요없긴 하지만 @LeeOhHyung 님이 쓰신 예제에서도 써져있고, 추후에 쓸 일이 생길거라 생각해서 데이터 클래스는 공통으로 다 쓸려고 생각하고 있었는데, 합의점을 찾아봐야 할 부분 같아요.

@SerializedName은 Gson에서 직렬화 할 때 변수이름을 저걸로 변환해주는 건데 대신 쓰는 건 아니에요.

https://simsi6.tistory.com/69

https://woowabros.github.io/experience/2017/10/17/java-serialize.html

LeeOhHyung replied:
추가적으로 Intent 혹은 Bundle 객체에 데이터를 put, get 할때 Primitive Type or List Type 이외에 개발자가 만든 클래스를 put, get 해야 하는 상황이 있습니다. 그럴때 Serializable or Parcelable을 사용해주어야 합니다.

teddy-choe replied:
자세한 설명 감사드립니다~~오형님도 추가 설명 감사용

teddy-choe commented on app/src/main/java/kr/yapp/teamplay/teammain/presentation/AdapterBindings.kt:
```diff
@@ -0,0 +1,11 @@
+import androidx.recyclerview.widget.RecyclerView
+import kr.yapp.teamplay.teammain.data.TeamMainFeedItem
+
+@BindingAdapter("bind:item")
```
BindingAdapter의 역할이 무엇인지 설명해주실 수 있을까요??

ohjuntaek replied:
리사이클러뷰에서 사용하는 데이터를 바인딩시키기 위해 사용해습니다.

https://medium.com/@futureofdev/%EC%95%88%EB%93%9C%EB%A1%9C%EC%9D%B4%EB%93%9C-what-is-databinding-recyclerview-e67abb855788

teddy-choe replied:
감사합니다~~링크걸어주신 포스팅도 좋네요!

teddy-choe reviewed (commented):
수고하셨습니다. 역시 코틀린 코드를 잘 사용하시는 것 같네요! 많이 배우고갑니다

LeeOhHyung commented on app/src/main/java/kr/yapp/teamplay/teammain/presentation/TeamMainAdapter.kt:
```diff
@@ -0,0 +1,105 @@
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
+        when (viewType) {
+            0 -> {
+                ResultTypeViewHolder(
+                    DataBindingUtil.inflate(
+                        LayoutInflater.from(parent.context),
+                        R.layout.rv_item_team_main_result_type,
+                        parent,
+                        false
+                    )
+                )
+            }
+            1 -> {
+                NoticeTypeViewHolder(
+                    DataBindingUtil.inflate(
+                        LayoutInflater.from(parent.context),
+                        R.layout.rv_item_team_main_notice_type,
+                        parent,
+                        false
+                    )
+                )
+            }
```
`ViewType`을 `enum`이나 `static const val`로 관리하는건 어떻게 생각하시나요??

ohjuntaek replied:
추후 enum으로 리팩토링 작업하겠습니다

LeeOhHyung commented on app/src/main/java/kr/yapp/teamplay/teammain/presentation/TeamMainAdapter.kt:
```diff
@@ -0,0 +1,105 @@
+            if (binding.resultItem!!.isWin) {
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+                    binding.resultContent = Html.fromHtml(
+                        String.format(
+                            "<b>%s</b>과의 경기에서 승리하셨습니다.",
+                            resultItem?.teamName
+                        ), Html.FROM_HTML_MODE_COMPACT
```
String.format을 바로 사용하면 될듯한 느낌인데, HTML 문법으로 사용하신 배경이 따로 있을까요??

ohjuntaek replied:
볼드체를 부분 적용하기위해 사용했어요

LeeOhHyung replied:
Spannable or SpannableString 을 사용해보는것은 어떨까요??

ohjuntaek replied:
오 이렇게 좋은게.... 감사합니다 추후 변경할게요

LeeOhHyung commented on app/src/main/res/drawable/item_notice_type_btn_label.xml:
```diff
@@ -0,0 +1,11 @@
+    <stroke
+        android:width="1dp"
+        android:color="#fe7346" />
+
+    <corners android:radius="30dp" />
+
```
실행은 안시켜보았는데, `radius` 30dp 으로 주면 짜부되지 않던가요???? (저는 짜부가 되어버려서 ..)

ohjuntaek replied:
음 제 폰(갤럭시 s8)에서 보기엔 괜찮은거 같은데 다음에 같이 확인해봐야겠네요.

LeeOhHyung commented on app/src/main/res/layout/rv_item_team_main_result_type.xml:
```diff
@@ -0,0 +1,137 @@
+            name="result"
+            type="String" />
+    </data>
+
+    <androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+        android:layout_width="match_parent"
```
여기 `xmlns`도 최상단 태그로 이동해야 할듯 해요

LeeOhHyung reviewed (commented):
고생하셨습니다.
하나만 제안드리자면, 저희 PR 제목도 너무 제각각이라 PR제목 컨벤션도 정해볼 필요가 있을듯 합니다.
코멘트 확인 부탁드려요~

ohjuntaek committed:
FEAT : TeamMainFeed 서버 연동

ohjuntaek committed:
FEAT : TeamMainFeed 서버 연동 및 merge conflict 해결 및 shimmer 추가

- 이전에 fragment navigation 예제로 썼던 MainActivity, TeamFeedFragment 삭제
- gradle에 disable: SpUsage 추가
- teammain 패키지 각각의 계층 패키지로 옮김
- Usecase 추가

- shipmmer 추가
- retrofit onFail 시 오류 시 토스트 메세지 출력

teddy-choe commented on app/src/main/java/kr/yapp/teamplay/domain/entity/teammain/TeamMainItemType.kt:
```diff
@@ -0,0 +1,5 @@
+package kr.yapp.teamplay.domain.entity.teammain
+
+enum class TeamMainItemType {
+    NOTICE, MATCH_RESULT
```
여기서 enum을 사용한 이유가 궁금합니다. 어댑터에서 뷰타입을 정하기 위함인 것 같은데 enum을 사용하는 이유가 무엇인지 알 수 있을까요??

ohjuntaek replied:
가독성과 유지보수를 위해 이넘을 사용했어요

teddy-choe reviewed (approved):
수고하셨습니다! 특히 리사이클러뷰 뷰타입을 나눈 부분에서 많이 배웠습니다.

ohjuntaek committed:
Squashed commit of the following:

commit c2d825b1e81055d0690ab8de780928d8d598dfee
Author: ChangikChoe <46737716+ckddlr0610@users.noreply.github.com>
Date:   Wed May 20 23:56:35 2020 +0900

    [PR-39] MatchList, MatchDetail 뷰 작업 (#42)

    * XML 작업

    * Match List 더미 데이터 작업 완료

    * Match List, Match Detail View 작업

    * Signin, Signup 일부 버그 HotFix, 엑티비티 로직 프래그먼트로 분리
